### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by Kibouo**
PR for #291

- Currently adding error messages based on context. However, I'm not familiar with the code base so **thorough review is highly recommended** to improve error messages
- ... especially for the `fzf` and `core` modules.

TODO:
- [x] Every call to `bash` now has an error message similar to "Failed to execute bash". This should probably be related to the actual command ran, i.e. params given to `bash`.
- [x] A lot of `anyhow::[with_]context()` is used to add, well, context. It might be better to create custom error variants for repetitive errors.
- [x] `navi fn welcome` is broken.
- [x] Add "file a bug report" message to top-level error.
- [x] Test for potential breakages. Logic changes have been avoided as much as possible but stuff happens.
- [x] `navi repo browse` and `navi repo add https://github.com/denisidoro/navi` crash. Root cause is `fzf` not returning data due to window layout error.
image
- [x] bump version

### CODE REVIEW

**Comment by Kibouo** [comment]
My approach currently is to:
- eliminate all `unwrap()`, except in tests
- `expect()` when calling an internal function with highly controlled input, since these "ugly" panic messages ~will~ should only ever be seen by devs.
- return as much `Result` as possible, as the added context to each layer of errors will be nicely formatted and hopefully useful for the end-user.

**Comment by denisidoro** [comment]
The `navi repo` bug is fixed in master 👍

**Comment by Kibouo** [comment]
Kindly re-review changes. Please check whether the actual messages are clear.

Also added a wrapping error telling the end-user to file an issue. Ex:
image

This required the error type `FileAnIssue` to be made public in the library. It had to be used in the bin. `handle_config` is called multiple times so it would look weird to potentially have multiple of these messages nested. Is this change ok?

A few things remain to be done as can be seen in the TODO of the first post.

**Comment by Kibouo** [comment]
Would this be a bump in version to 2.4.0?

File: src/flows/repo.rs [hunk 1]
```diff
@@ -2,25 +2,28 @@ use crate::filesystem;
...
-    let repo_path_str = format!("{}/featured", filesystem::tmp_path_str());
+pub fn browse() -> Result<(), Error> {
+    let repo_path_str = format!("{}/featured", filesystem::tmp_path_str()?);
 
-    filesystem::remove_dir(&repo_path_str);
-    filesystem::create_dir(&repo_path_str);
+    // The dir might not exist which would throw an error. But here we don't care about that
+    // since our purpose is to make space for a new directory. Potential other errors
+    // (e.g due to lack of permission) will cause an error during dir creation.
+    let _ = filesystem::remove_dir(&repo_path_str);
```
Comment (denisidoro) [inline src/flows/repo.rs]: Why using `let _ =`?

File: src/flows/repo.rs [hunk 2]
```diff
@@ -2,25 +2,28 @@ use crate::filesystem;
...
 use walkdir::WalkDir;
 
-pub fn browse() -> Result<(), Box<dyn Error>> {
-    let repo_path_str = format!("{}/featured", filesystem::tmp_path_str());
+pub fn browse() -> Result<(), Error> {
+    let repo_path_str = format!("{}/featured", filesystem::tmp_path_str()?);
 
-    filesystem::remove_dir(&repo_path_str);
-    filesystem::create_dir(&repo_path_str);
+    // The dir might not exist which would throw an error. But here we don't care about that
```
Comment (denisidoro) [inline src/flows/repo.rs]: Do you suggest another approach? I agree this is very error-prone

File: src/handler.rs [hunk 3]
```diff
@@ -2,21 +2,40 @@ use crate::flows;
...
         None => flows::core::main(Variant::Core, config, true),
+
         Some(c) => match c {
             Preview { line } => flows::preview::main(&line[..]),
-            Query { query } => flows::query::main(query.clone(), config),
-            Best { query, args } => flows::best::main(query.clone(), args.to_vec(), config),
-            Search { query } => flows::search::main(query.clone(), config),
+
+            Query { query } => {
+                let error_string = format!("Failed to filter cheatsheets for {}", &query);
```
Comment (denisidoro) [inline src/handler.rs]: Wouldn't `with_context` prevent us from generating this str 100% of the times?

File: src/parser.rs [hunk 4]
```diff
@@ -3,38 +3,80 @@ use crate::filesystem;
...
+                    "--preview-window" => opts.preview_window = Some(value.to_string()),
+                    "--header" => opts.header = Some(value.to_string()),
+                    "--overrides" => opts.overrides = Some(value.to_string()),
+                    _ => (),
+                }
+                Ok(())
+            } else if let [flag] = flag_and_value {
+                Err(anyhow!("No value provided for the flag `{}`", flag))
+            } else {
+                unreachable!() // Chunking by 2 allows only for tuples of 1 or 2 items...
```
Comment (denisidoro) [inline src/parser.rs]: Why not use an else in the line 72 instead?

File: src/structures/error/file_issue.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+use std::fmt::Debug;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+#[error(
+    "\rHey listen! Navi encountered a problem.
```
Comment (denisidoro) [inline src/structures/error/file_issue.rs]: 😂

File: src/flows/repo.rs (hunk 1 above)
Comment (Kibouo) [reply src/flows/repo.rs]: Rust requires you to use the return value of `remove_dir`. But as stated in the comment above it, we don't care about the potential `Err` value. And the `Ok` type is `()` anyway.

File: src/parser.rs (hunk 4 above)
Comment (Kibouo) [reply src/parser.rs]: The `chunks(2)` call on :46 returns slices (`&[..]`) of the collection. The problem is that, despite telling `chunks` that the slices should be `2` long, it doesn't know this at compile time. So Rust wants us to handle every possible case of slice length exhaustively.

Concrete this means that Rust wants the following let binding to be exhaustive.
```rust
} else {
    let [flag] = flag_and_value;
    Err(anyhow!("No value provided for the flag `{}`", flag))
}
```
image

You could do the following but obviously it wouldn't be as helpful an error:
```rust
} else {
    Err(anyhow!("No value provided for a flag"))
}
```

File: src/flows/repo.rs (hunk 2 above)
Comment (Kibouo) [reply src/flows/repo.rs]: Checking whether the dir exists before trying to remove it is a bit more robust, but not worth it imho. The check + remove is not atomic so `remove_dir` would still be able to return the same errors. Just less frequently.

File: src/handler.rs (hunk 3 above)
Comment (Kibouo) [reply src/handler.rs]: `query` is part of `config`. But `config` is given away to `flows::query::main` on :17, which makes it impossible to use `query` after that. So we need to make a copy of `query` before that happens. 

`let query_clone = query.clone()` and then passing that to `with_context` would be a possibility. Not sure how smart the compiler is. I'll change it to be sure :thinking:

File: src/filesystem.rs [hunk 6]
```diff
@@ -1,90 +1,122 @@
...
-    P: AsRef<Path>,
+    P: AsRef<Path> + Display,
 {
+    let error_string = format!("Failed to read lines from `{}`", filename);
     let file = File::open(filename)?;
-    Ok(io::BufReader::new(file).lines())
+    io::BufReader::new(file)
+        .lines()
+        .map(|line| line.map_err(Error::from))
+        .collect::<Result<Vec<String>, Error>>()
```
[the remark this answers is not in the document]
Comment (Kibouo) [reply src/filesystem.rs]: [In the `Err` case it short circuits.](https://doc.rust-lang.org/src/core/result.rs.html#1438) But best case it indeed allocates unnecessarily. Fixed!

**Pull request merged**